Override default proxy from provider

juliancheal opened the pull request:
Initial stab at allowing providers to override the default proxy.

Can allow for situation where all trafic goes through one proxy,
except for example the Amazon traffic goes through another.

juliancheal commented on config/settings.yml:
```diff
@@ -902,11 +902,16 @@
-  :password:
-  :port:
-  :user:
+ec2_http_proxy:
```
Didn't mean to commit these changes to the `settings.yml`, but works as a good example of how I imagine overriding the proxy would work.

juliancheal replied:
I thought we could namespace the proxies by `ems.ems_type` i.e. `ec2`, `openstack`.

What was the thought of how this is populated… by a form on the app, or by hand by the end user.

If the latter we'd need to make sure they knew how to format it. We could just auto create the fields with empty values, like we do for `http_proxy`

blomquisg commented:
I like the mixin idea for this!

@Fryguy, @bronaghs, any thoughts on this?

bronaghs commented:
very nice @juliancheal  👍

juliancheal commented on app/models/manageiq/providers/azure/manager_mixin.rb:
```diff
@@ -27,7 +27,7 @@ def verify_credentials(_auth_type = nil, options = {})
   module ClassMethods
     def raw_connect(client_id, client_key, azure_tenant_id, subscription, proxy_uri = nil)
-      proxy_uri ||= VMDB::Util.http_proxy_uri
+      proxy_uri ||= http_proxy_uri
```
```ruby
     NameError:
       undefined local variable or method `http_proxy_uri' for #<Class:0x007fb5e6cd3310>
     # /Users/Jools/.rvm/gems/ruby-2.2.4/gems/activerecord-5.0.0/lib/active_record/dynamic_matchers.rb:21:in `method_missing'
     # ./app/models/manageiq/providers/azure/manager_mixin.rb:32:in `raw_connect'
     # ./app/models/manageiq/providers/azure/manager_mixin.rb:64:in `discover'
     # ./spec/models/manageiq/providers/azure/cloud_manager_spec.rb:97:in `block in recorded_discover'
```

djberg96 commented:
👍 Looking good.

juliancheal commented on app/models/manageiq/providers/azure/manager_mixin.rb:
```diff
@@ -6,7 +6,7 @@ def connect(options = {})
     client_id  = options[:user] || authentication_userid(options[:auth_type])
     client_key = options[:pass] || authentication_password(options[:auth_type])
-    self.class.raw_connect(client_id, client_key, azure_tenant_id, subscription, options[:proxy_uri])
+    self.class.raw_connect(client_id, client_key, azure_tenant_id, subscription, options[:proxy_uri] ||= http_proxy_uri)
```
NB. @juliancheal change this to `||`

juliancheal commented:
Thinking about where to put the additional proxy settings. Should we create empty values, so users can just add the data. As the naming is dependent on the value returned from `ems.emstype`

Second should we put those extra values in `settings.yml` or include an additional settings file for proxy info? /cc @blomquisg

juliancheal commented:
Google Proxy support is waiting on https://github.com/fog/fog-google/issues/153

durandom commented:
> I like the mixin idea for this!

I dont want to be the :tada: :hankey: - but I dont get why we would need a a mixin for this? 
Wouldnt it be enough to have it as a method in `ExtManagementSystem` or even better in `BaseManager` ?

durandom commented on app/models/mixins/provider_proxy_mixin.rb:
```diff
@@ -0,0 +1,13 @@
+  extend ActiveSupport::Concern
+
+  def http_proxy_uri
+    proxy = VMDB::Util.http_proxy_uri("#{emstype}_http_proxy".to_sym)
```
How about following the suffix scheme like refresh and others
https://github.com/ManageIQ/manageiq/blob/faec76397e8e77d65e98432d8d46e38357632832/config/settings.yml#L1100

also I would go with the `defaults` like [here](https://github.com/ManageIQ/manageiq/blob/faec76397e8e77d65e98432d8d46e38357632832/config/settings.yml#L1101)

but I'm not sure about the implications of changing the structure of settings? @blomquisg ?

durandom replied:
Or something like the refresher:
https://github.com/durandom/manageiq/blob/c445fb3fd2b40e3e2a3377e2ed78bb20f12589ed/app/models/manageiq/providers/base_manager/refresher.rb#L22

juliancheal replied:
Sorry @durandom not sure what you're saying here :(

durandom replied:
yeah, sorry, my late night brain dump reads a bit unclear :smile: 

In settings.yml we have currently 2 ways for provider specific settings.

* suffix: like `event_catcher_vmware` 
* nested: like `ems_refresh` -> `ec2`

This would introduce a third, prefix: `vmware_http_proxy`

I would go with the nested approach and move that logic to `VMDB::Util.http_proxy_uri` 

```ruby
VMDB::Util.http_proxy_uri(ems_type)

def self.http_proxy_uri(proxy_config = nil)
  proxy = Settings.http_proxy[subsystem] || Settings.http_proxy
```

Then we can have in settings.yml
```yaml
:http_proxy:
  :host:
  :password:
  :port:
  :user:
  :ec2:
     :host:
     :password:
     :port:
     :user:
```

It would be even nicer with defaults like [here](https://github.com/ManageIQ/manageiq/blob/faec76397e8e77d65e98432d8d46e38357632832/config/settings.yml#L1101) - but I guess that would break backward compatability

juliancheal replied:
I'm with you, yeah that makes sense. :)

juliancheal commented:
> I dont want to be the 🎉 💩 - but I dont get why we would need a a mixin for this? 
> Wouldnt it be enough to have it as a method in ExtManagementSystem or even better in BaseManager ?

My original thinking was it might be more complex than just one method, and I thought a Mixin might encapsulate that nicely.

Do you think it would be better in BaseManager, or are you just avoiding more included Mixins? 😜

durandom commented:
> Do you think it would be better in BaseManager

I think so, because we are moving to get rid of ExtManagementSystem.

> or are you just avoiding more included Mixins? 

+1 if the mixin is only included in one class (including subclasses).
If its used in other classes too, then a mixin is ok, although probably a new class as a collaborator would be even better

juliancheal commented:
@blomquisg Travis is all green. I've updated the PR with @durandom's suggestions. I've added new values into `settings.yml` if that's all ok, then this PR should be good to go.

blomquisg commented on lib/vmdb/util.rb:
```diff
@@ -1,7 +1,10 @@
+      # TODO: (julian) This looks really messy. Need to make it look nicer.
+      vmdb_proxy = VMDB::Config.new("vmdb").config[:http_proxy]
+      return nil unless vmdb_proxy
+      proxy = vmdb_proxy[proxy_config] || {}
```
So, it looks like if this were on an upgraded appliance with "legacy" configs (i.e., no `:default` under `:http_proxy`, just the proxy settings themselves) then the worst case is that the proxy settings would be empty.

I think you should log a warning here saying that you couldn't find any proxy settings for the given `proxy_config`.  That way, legacy files will at least warn the user that they need to update their proxy configs.

durandom replied:
Actually I would make this backward compatible unless, @Fryguy do we have tools in place to migrate settings?

To make it backward compatible, this should work:
```ruby
proxy = vmdb_proxy[proxy_config] || vmbd_proxy
```

juliancheal replied:
> So, it looks like if this were on an upgraded appliance with "legacy" configs (i.e., no :default under :http_proxy, just the proxy settings themselves) then the worst case is that the proxy settings would be empty.

Ah yeah, that's a good idea. I forgot about legacy appliances. :(

juliancheal replied:
How about:

```bash
WARN -- : Could not find proxy setting for default
```

and then for each provider

```bash
WARN -- : Could not find proxy setting for azure
```

blomquisg commented:
> I've updated the PR with @durandom's suggestions

Ah, I thought @durandom was suggesting to just put the `http_proxy_uri` method in `BaseManager`.

durandom commented on lib/vmdb/util.rb:
```diff
@@ -1,7 +1,10 @@
-      proxy = VMDB::Config.new("vmdb").config[:http_proxy] || {}
+    def self.http_proxy_uri(proxy_config = :default)
+      # TODO: (julian) This looks really messy. Need to make it look nicer.
+      vmdb_proxy = VMDB::Config.new("vmdb").config[:http_proxy]
```
@Fryguy cant this be just ?
```ruby
vmdb_proxy = Settings.http_proxy
```

blomquisg replied:
@juliancheal can you update this one?  `VMDB::Config.new("vmdb").config` ~= `Settings`.

juliancheal replied:
@blomquisg not sure what you mean?

durandom replied:
This line should be 
```
vmdb_proxy = Settings.http_proxy
```

juliancheal replied:
The problem with `vmdb_proxy = Settings.http_proxy` is the tests no longer pass. https://github.com/ManageIQ/manageiq/pull/10368/commits/73126d9c379b5c82f38977a80b336f08a59baf00

As in the tests we pass in a new options hash, which when the code is `VMDB::Config.new("vmdv").config[:http_proxy]` works great, but with ` Settings.http_proxy` we can't stub the `settings.yml` that way.

durandom commented on config/settings.yml:
```diff
@@ -903,10 +903,21 @@
+    :password:
+    :port:
+    :user:
+  :amazon:
```
Just realized in another PR, that this is `ec2` or `ec2_network`. See [here](https://github.com/ManageIQ/manageiq-providers-amazon/search?utf8=%E2%9C%93&q=ems_type&type=Code)

juliancheal commented:
> Just realized in another PR, that this is ec2 or ec2_network. See here

Good spot. What's the difference between `ec2` & `ec2_network`?

durandom commented:
> What's the difference between ec2 & ec2_network?

The first one is the CloudManager and the other the NetworkManager.

juliancheal commented:
> The first one is the CloudManager and the other the NetworkManager.

Presumably we'd have to set both of them then? This is getting complicated 😜

juliancheal commented:
> Ah, I thought @durandom was suggesting to just put the http_proxy_uri method in BaseManager.

@durandom had lots of suggestions 😜

durandom commented:
> Presumably we'd have to set both of them then? This is getting complicated

If you branch on `ems_type`, yes. Actually this whole thing is a perfect use case for the `Provider` class.

All Managers should belong to a Provider. And then you could have a chain of Fallbacks:
Default -> Provider -> Manager

And that all just being the classes :smile: You could even have special settings at the instance level...

Dont know for sure. I guess the intuitive thing would be Default -> Provider
But then, we dont have a way, yet, to get the Provider  name, haha

durandom commented:
My main unaddressed concern is backwards compatibility. Could we find out how we deal with this? Is there a migration path, like a script we run on upgrades? Or should we just make it backwards compatible

```ruby
vmdb_proxy = Settings.http_proxy
# use pre euwe config at http_proxy level for backwards compatibility
proxy = vmdb_proxy[proxy_config] || vmdb_proxy
```

If we chose the latter path I'd add the code comment about backward compatibility and add some tests that nobody breaks it

djberg96 commented:
@durandom Can't the default value with a singleton method be used for backwards compatibility?

durandom commented:
> Can't the default value with a singleton method be used for backwards compatibility?

@djberg96 you mean `Settings.http_proxy[:default]`? 
Currently the proxy conf is at `Settings.http_proxy.host` etc, so this has to used for `settings.yml <= darga`

blomquisg commented:
I'm `wip`ping this until [Marcel's comment](https://github.com/ManageIQ/manageiq/pull/10368#issuecomment-243379131) is addressed.

durandom commented on app/models/manageiq/providers/base_manager.rb:
```diff
@@ -17,5 +17,15 @@ def ext_management_system
       self.class::Refresher
     end
+
+    def http_proxy_uri
```
minor but this method could be a oneliner

```ruby
def http_proxy_uri
  VMDB::Util.http_proxy_uri(emstype) || VMDB::Util.http_proxy_uri
end
```

durandom commented on app/models/manageiq/providers/google/manager_mixin.rb:
```diff
@@ -26,23 +26,24 @@ def connect(options = {})
         :google_json_key_string => google_json_key,
         :app_name               => I18n.t("product.name"),
-        :app_version            => Vmdb::Appliance.VERSION,
+        :app_version            => Vmdb::Appliance.VERSION
```
I'd leave the trailing comma

durandom commented on lib/vmdb/util.rb:
```diff
@@ -1,7 +1,17 @@
+        $log.warn("Could not find proxy setting for #{proxy_config}")
+      end
+
+      proxy = vmdb_proxy[proxy_config] || {}
```
If this line is changed to
```ruby
proxy = vmdb_proxy[proxy_config] || vmdb_proxy
```

you make sure that former configurations still work

durandom commented on spec/lib/vmdb/util_spec.rb:
```diff
@@ -6,46 +6,46 @@
     end
 
     it "without a host" do
-      stub_server_configuration(:http_proxy => {})
```
Please add a spec for testing the backward compatible settings. In case the customer did not start with a new settings file. Naming it like 'it "returns proxy for old settings"'

durandom commented:
I have no idea how deal with upgrading settings. So I reviewed under the premise that a settings.yml and its db counterpart is merged upon upgrade - the db one come from the import of the current db and the file comes from the appliance and then it's merged. If thats the case we have to take care of the old proxy configuration still present at `Settings.http_proxy` vs the new one at `Settings.http_proxy.default`

durandom commented on lib/vmdb/util.rb:
```diff
@@ -1,7 +1,17 @@
+        $log.warn("Could not find proxy setting for #{proxy_config}")
+      end
+
+      proxy = vmdb_proxy[proxy_config].to_h || vmdb_proxy.to_h
```
This wont work. `vmdb_proxy[proxy_config].to_h` is always a hash. And the `.to_h` is not needed because you can access `Settings` like a hash and with `.methods`. Try it in `rails c`, it's great, like an OpenStruct

durandom commented:
@juliancheal well done and thanks for bearing with me ;)

Do you also do that change on the amazon PR?

juliancheal commented:
> @blomquisg merge?

Should I squash some of these commits?

juliancheal commented:
> Do you also do that change on the amazon PR?

You'll have to remind me, what change?

juliancheal commented:
@miq-bot remove_label wip

chessbyte commented:
@blomquisg This did not make it into Euwe branch.  Please label `euwe/yes` or `euwe/no`, as appropriate.

durandom commented:
@miq-bot add_label euwe/yes

chessbyte commented:
Euwe Backport details:
```sh
$ git log
commit 4a31d98f4afcacac30c637c72f01a2d2cff9e9ba
Author: Greg Blomquist <blomquisg@gmail.com>
Date:   Fri Sep 23 10:41:42 2016 -0400

    Merge pull request #10368 from juliancheal/provider_proxy
    
    Override default proxy from provider
    (cherry picked from commit a27866cd11e6eaac923b857501e3efc94ee8d061)
```
